### ISSUE TITLE
idea - should the postgres copy path get a copy_insert/create method?

### ISSUE THREAD

**Issue by jerch** [netzkolchose/django-fast-update#4]
There is a possibility for a much faster `copy_create` alternative to `bulk_create`. From some initial tests that would be 10-20 times faster than the current implementation, but will have several restrictions/limitations - esp. the ID returning feature is hard to get done right (there might be workarounds possible from transaction id, not sure yet if thats a viable path).

**Comment by jerch** [comment]
What also would work - copy the rows into a temp table with a unique position column and do a normal insert with returning the pk, but ordered by that position column. Well would need an additional temp table step, hmm.

Edit:
This is actually much easier and the recommended way - get n nextval from the pk sequence and apply them to the objects before doing the copy step. Done. (+ drop/recreate index, all in a transaction and so on...).

This is guaranteed to work, as the sequence progression is not transactional, thus increments globally for all connections/transactions. If our own transaction fails, no data will enter the table, index is still intact (indexes are transactional), but the sequence will have gaps from the failed insert attempt, which is not an issue at all.

**Comment by jerch** [comment]
Some early numbers with the sequence trick:
```python
>>> Foo.objects.all().delete()
(1000000, {'exampleapp.Foo': 1000000})
>>> o=gen(1000000)
>>> _=timer(lambda:Foo.objects.bulk_create(o))
62.56771206855774
>>> Foo.objects.all().delete()
(1000000, {'exampleapp.Foo': 1000000})
>>> o=gen(1000000)
>>> _=timer(lambda:copy_create(Foo.objects.all(), o))
14.51438307762146
```
It is ~4x faster.

Without pk it is:
```python
>>> o=gen(1000000)
>>> _=timer(lambda:copy_create(Foo.objects.all(), o))
3.946061134338379
```
Wow, now this is ~15x faster. 😮

**Comment by jerch** [comment]
This is my best take on _copy_create_ so far:
```python
>>> o=gen(1000000)
>>> _=timer(lambda:Foo.objects.bulk_create(o))
63.24271035194397
>>> o=gen(1000000)
>>> Foo.objects.all().delete()
(1000000, {'exampleapp.Foo': 1000000})
>>> _=timer(lambda:copy_create(Foo.objects.all(), o))
5.775318145751953
```
which is ~11 times faster. Does not handle MT models yet. It returns a queryset covering the new items, directly patching the given objects is not possible with the current approach. Indeed might have to switch to a temp table and doing INSERT ... RETURNING from there (would also make MT table handling alot easier).

Also found a bug in the threaded code not propagating exceptions back to caller thread

Edit:
Seems proper MT handling is only possible with the temp table trick. Otherwise we lose the order on the first table already and cannot match against the objs anymore. Hmm.

**Comment by jerch** [comment]
Hmm MT is really tricky with COPY FROM and needs alot impl before we can test. Basic idea:
- group fields to their models/tables first
- patch all objs to contain an incrementing copy_id attribute
- push basemodel field + copy_id to first temp table
- do a table insert from 1st temp table ordered by copy_id returning pks
- patch new pks on all objs
- push remaining fields to second temp table
- iterate over all model/field groups:
  - do a table insert from 2nd temp table

Unclear, which way is better, also hard to test due to high impl needs:
1. push all data fields right into 1st temp table (one big COPY FROM), load pks into 2nd temp table (smaller COPY FROM), update 1st temp table from 2nd, update target tables from now fully populated 1st temp table
2. push only basemodel fields to 1st temp table (small COPY FROM), after pk patching on python side load remaining fields into 2nd temp table (much bigger COPY FROM due to retransferring pks), update target tables from 2nd temp table

The 2nd way seems more straight-forward to implement, but the retransfer of the pks might just make it alot worse. The 1st way only needs a tiny 2nd temp table with (pk, copy_id) columns. But now we repeat the pk merging on database side. Hmm.

Edit:
This gets further complicated by objs with and w'o prepolulated pk. Which leads to different table insert on the basemodel.


Edit2:
Maybe CTE for the rescue here? Could write the returning pks directly into the 1st temp table updating all pk fields. Avoids any 2nd COPY FROM transfer, no 2nd temp table, and we can query the temp at the end to finally patch the objs. Or with less of a perf burden - return a queryset from the pks.

**Comment by jerch** [comment]
Just finished my first hacky CTE-based version of _copy_create_ with MT support and wanted to do some perf comparisons - turns out _bulk_create_ does not support it at all? 😮  Oh well, here are the numbers compared to loop-saving:
```python
>>> o_mt=gen_mt(100000)
>>> _=timer(save_omt)
135.360604763031
>>> o_mt=gen_mt(100000)
>>> _=timer(lambda:copy_create(MultiSub.objects.all(), o_mt))
4.7425596714019775
```
which is ~28 times faster. It still lacks a few pk details, so the speed benefit might shrink to 15-20x.

Now it takes longer to delete than to insert them:
```python
>>> o_mt=gen_mt(1000000)
>>> _=timer(lambda:copy_create(MultiSub.objects.all(), o_mt))
36.41003084182739
>>> timer(lambda:MultiSub.objects.all().delete())
116.68937063217163
(2000000, {'exampleapp.MultiSub': 1000000, 'exampleapp.MultiBase': 1000000})
```

**Comment by jerch** [comment]
Comparing the MT impl with the first one on a non-MT model - it is 5x slower. It seems that the temp tables are not a good idea at all. Also the CTE is quite heavy (half of the total runtime).

Problem with MT is that we need an exact pk to obj matching to not mess up field values. For COPY FROM postgres gives no insert order promise at all, thus we need at least one temp table to get RETURNING ordered by our custom copy_id. But we can prolly safe data copies by doing all later tables with COPY FROM directly and with the pks applied to the o2o fields.

**Comment by jerch** [comment]
Ofc the runtime behavior is not as simple as doubling from the CTE usage - the critical paths are the inserts done on the target tables + the pk updates on the interim temp table (for 100k rows):
- my first non-MT approach does the COPY FROM directly into the target table with one index on pk --> ~0.6s
- the MT approach does COPY FROM to a temp table (~0.5s) + INSERT ... RETURNING to the target table --> ~1.2s for a non MT model
- for MT model things get worse:
  - COPY FROM to a temp table ~0.4s
  - CTE INSERT ... RETURNING + UPDATE pk fields ~0.7s
  - INSERT multisub ~0.3s
  - total (including transaction commit) ~2.9s

The weird part for MT - the single positions dont add up to the total runtime, seems postgres postponed alot of work to COMMIT here?

**Comment by jerch** [comment]
From here on it is a design question, whether to make things more in line with _bulk_create_ or not:
- My first approach is overwhelmingly fast, but cannot patch the given objs for their new pks. It also cannot be used with MT and cannot serve the other arguments of _bulk_create_, as it misses the INSERT step for ON CONFLICT handling.
- The second MT approach can patch the objs, and in theory serve the other arguments of _bulk_update_. It runs much slower and works for MT models.

No clue yet, what is better - to be more in line with the behavior of the default impl is always a good thing, but the de-optimization introduced by the exact obj to pk matching at least doubles the runtime. Maybe clever argument eval can preserve the fast path for non-MT models (might need an explicit _returns_ argument though)?

**Comment by jerch** [comment]
About the timings not adding up in MT:

Thats what is needed for 1M multisub instances:
```
startup: 0.5521700382232666
COPY: 3.7627549171447754
ANALYZE: 0.1126248836517334
CTE: 7.534581661224365
INSERT <class 'exampleapp.models.MultiSub'> 3.8671469688415527
UNDER TRANSACTION: 15.367772817611694
28.6779682636261
```
where COPY + ANALYZE + CTE + INSERTs add up nicely to UNDER TRANSACTION, but missing ~14s of the total runtime.

Doing the same without a transaction reveals this:
```
startup: 0.5268363952636719
COPY: 3.7841379642486572
ANALYZE: 0.13716912269592285
CTE: 7.382129669189453
INSERT <class 'exampleapp.models.MultiSub'> 16.3602454662323
UNDER TRANSACTION: 27.698508977890015
28.251802444458008
```
wow, seems the sub table constraints (on o2o?) cause the huge additional runtime, which hides behind COMMIT with a transaction, hmm. Technically we already know that our additions fullfill the constraint here (as we pulled the pk values over), maybe a temporary lift of the constraint can help here?

Edit:
The constraints are:
```
Indexes:
    "exampleapp_multisub_pkey" PRIMARY KEY, btree (multibase_ptr_id)
Foreign-key constraints:
    "exampleapp_multisub_multibase_ptr_id_51f1267c_fk_exampleap" FOREIGN KEY (multibase_ptr_id) REFERENCES exampleapp_multibase(id) DEFERRABLE INITIALLY DEFERRED
```
The constraint on `multibase_ptr_id` with the `DEFERRABLE INITIALLY DEFERRED` moves its workload to COMMIT.

Edit2:
Lifting the constraints temporarily helps:
```
startup: 0.5664236545562744
COPY: 3.9924724102020264
ANALYZE: 0.10381770133972168
CTE: 7.447082757949829
INSERT <class 'exampleapp.models.MultiSub'> 3.7418081760406494
IN TRANSACTION: 15.294859170913696
15.920311450958252
```
(to see the difference simply done with a DISABLE/ENABLE TRIGGER ALL, needs to be lifted on the o2o field only)

**Comment by jerch** [comment]
With a few more tiny optimizations (like not explicitly updating the o2o fields on the temp table, instead pulling directly from the pk field) I get the runtime of the MT case down to (n=1M):
```python
>>> o_mt=timer(lambda:gen_mt(1000000))
11.803561449050903
>>> _=timer(lambda:copy_create(MultiSub.objects.all(), o_mt))
startup: 0.5265181064605713
COPY: 3.79563045501709
ANALYZE: 0.11489748954772949
CTE: 7.33696985244751
INSERT <class 'exampleapp.models.MultiSub'> 3.6288609504699707
IN TRANSACTION: 14.881783246994019
15.48543405532837
```
Now we are almost as fast the object creation itself. What still bugs me it the high runtime of the CTE and the INSERT on the subtable, which is still much higher than for a single table.

On a sidenote - why does it take like forever to delete things again?
```python
>>> timer(lambda: MultiSub.objects.all().delete())
115.15562582015991
(2000000, {'exampleapp.MultiSub': 1000000, 'exampleapp.MultiBase': 1000000})
```

**Comment by jerch** [comment]
For proper saving of MT instances we also have to take the pk precedence into account, which seems to be

id (real pk field) < pk (attribute) < any_o2o_ptr

The fact that the pk attribute beats any o2o field but not the real pk field seems slightly odd, but can easily be shimmed. Still need to check in which order multiple o2o fields will be applied...

**Comment by jerch** [comment]
Surprise - doing MT inserts not all in one is actually faster than trying to tackle it with the CTE (~10% faster). This results from the fact, that the indirection with a temp table is not needed for sub tables. The CTE approach has another drawback - it currently always assumes, that an MT insert would have to create the parent instances as well, which is not true for many MT use cases.

A better approach is to go back the "operate on a single model" idea, thus when given an MT model instances must have a pk already (o2o field). This simplifies the code alot and makes it possible to add to the subtable only.

**Comment by jerch** [comment]
With [803b5a1](https://github.com/netzkolchose/django-fast-update/pull/28/commits/803b5a16dce4e9ca79766ee8e3282b367aca8fdf) MT handling is simplified as follows:
- use `copy_create` to insert submodel instance fields
- use `copy_create_MT` for fully new MT instances, which deals with the parent creation by calling `copy_create` for each parent

The function now also supports dictionaries as object data with the argument _from_dict=True_. This is esp. helpful when dealing with tons of input data, which often already comes as dicts and where the instance creation can get quite heavy itself.

For comparison:
```python
>>> from fast_update.copy import copy_create
>>> 
>>> def gen(n):
...   objs = []
...   for _ in range(n):
...     objs.append(Foo())
...   return objs
... 
>>> def gen_dicts(n):
...   objs = []
...   for _ in range(n):
...       objs.append({
...         'f1':1, 'f2':2, 'f3':3, 'f4':4, 'f5':5,
...         'f6':6, 'f7':7, 'f8':8, 'f9': 9, 'f10':10
...       })
...   return objs
... 
>>> def timer(f):
...   from time import time
...   start = time()
...   res = f()
...   print(time()-start)
...   return res
... 
>>> _=timer(lambda:Foo.objects.bulk_create(gen(1000000)))
71.08409094810486
>>> _=timer(lambda:copy_create(Foo.objects.all(), gen(1000000)))
13.670178174972534
>>> _=timer(lambda:copy_create(Foo.objects.all(), gen_dicts(1000000), from_dict=True))
4.916342496871948
>>> _=timer(lambda:gen(1000000))
9.036413192749023
>>> _=timer(lambda:gen_dicts(1000000))
0.5003395080566406
```
Yeah most of the time is spend for object creation, hmm.

**Comment by jerch** [comment]
Same comparison with the MT model:

```python
>>> from fast_update.copy import copy_create, copy_create_MT
>>> 
>>> def gen_mt(n):
...   objs = []
...   for _ in range(n):
...     objs.append(MultiSub(s1=100,s2=200,b1=300,b2=400))
...   return objs
... 
>>> def gen_mt_dicts(n):
...   objs = []
...   for _ in range(n):
...     objs.append({'s1': 100, 's2': 200, 'b1': 300, 'b2': 400})
...   return objs
... 
>>> def timer(f):
...   from time import time
...   start = time()
...   res = f()
...   print(time()-start)
...   return res
... 
>>> _=timer(lambda:copy_create_MT(MultiSub.objects.all(), gen_mt(1000000)))
37.02809000015259
>>> _=timer(lambda:copy_create_MT(MultiSub.objects.all(), gen_mt_dicts(1000000), from_dict=True))
23.04407525062561
>>> o_mt=timer(lambda:gen_mt(1000000))
10.95585823059082
>>> o_mt=timer(lambda:gen_mt_dicts(1000000))
0.32112669944763184
```
And to compare this to the currently only available MT-way in django:
```python
>>> def mt_create(objs):
...   from django.db.transaction import atomic
...   with atomic():
...     for o in objs:
...       o.save()
... 
>>> timer(lambda: mt_create(gen_mt(1000000)))
1289.828085899353
```
`copy_create_MT` is ~56 times faster when using dicts. 😮

**Comment by jerch** [comment]
Here a graph showing the runtime for creating x instances of the _Foo_ model (with _bulk_create_ normalized to 1):

<img width="607" height="313" alt="Image" src="https://github.com/user-attachments/assets/6c9525d1-2997-4c01-99f1-7ae90083284d" />

`returns=True` is way more expensive, esp. for high rowcount. Only strict object to pk matching is implemented for _returns=True_, which only works with the temp table overhead in between. I removed the less penalizing unordered pk-set as return value, as the mechanics to get that are somewhat unreliable. If someone does not care for strict order of returned pks, getting the unordered pks can be shimmed with the help of the xmin system column:
```python
# important: atomic must open a fresh transaction context, not just a safepoint
# also do not change the table in question by other actions within that transaction
with atomic():
    copy_create(model...)
    table = model._meta.db_table
    pk_column = model._meta.pk.column
    pks = list(model.objects.raw(
        f'SELECT {pk_column} FROM {table} WHERE xmin = pg_current_xact_id_if_assigned()::xid')
    )
```
For strictly autoincrementing pks it is even simpler and slightly faster with a filter:
```python
with atomic():
    last_id = model.objects.aggregate(models.Max('pk'))['pk__max']
    copy_create(model...)
    pks = list(model.objects.filter(pk__gt=last_id).values_list('pk', flat=True))
```
**NOTE** - never try to associate the pks returned by those shims to the provided objects, as postgres gives no insert order guarantees for COPY FROM. If you need 1:1 pk to object matching use _returns=True_ instead.

**Comment by jerch** [comment]
Quickly hacked comparison to psycopg3's enhanced copy support, again creating x _Foo_ instances (both with _returns=False_, in seconds):

|   | PG2 | PG3 | speedup  |
|---|---|---|---|
| 10  | 0.0092 | 0.0080 | 1.15 |
| 100  | 0.0114 | 0.0102 | 1.12  |
| 1000  | 0.0299 | 0.0273 | 1.10 |
| 10000  | 0.0946 | 0.0642 | 1.47 |
| 100000  |0.4753 | 0.3034 | 1.57 |

Seems for low counts the benefit is not that much, but things start to fly for +10k.

On my laptop PG2 saturates the CPU for ~55% while PG3 gets postgres to ~90% writing with ~13 MB/s (guess my CPU is already the bottleneck here...)

**Comment by jerch** [comment]
Something weird is going on with `get_db_prep_save` - runtime for 100k foos with PG3 fast copy:
 - no explicit value preparation - 0.32 s (baseline, we prolly cannot get it any faster than this)
 - own encoder - 0.41 s
 - with django method - 2.4 s

Needs some investigation, what makes the django's value adaption this expensive and whether we can omit that.

Edit: Found the issue - `get_db_prep_save` adapts the int fields to the `Int4` type, which costs alot of construction runtime. But since we use `copy.set_types` already, this is not needed with the copy-binary interface. Solution: Again provide own slim encoders for fields that need it explicitly like `Jsonb`. For all others go with the auto adaption from `copy.set_types`.

**Comment by jerch** [comment]
Since the new psycopg3 COPY/binary semantics are very promising performance- and typesafety-wise, I gonna do the transition for _copy_update_ also here. Already found a way to deal with the slightly off value adaption of `get_db_prep_save` for COPY/binary, but it is quite some impl work (test cases! test cases!).

The basic idea is the following - there are 4 preparation modes:
- 'none'
  No value preparation at all. Useful, if the values are guaranteed to be supported by psycopg's auto adaption (e.g. already validated by other means). Fastest mode and prolly preferrable on really huge datasets.
- 'strict'
  Does strict type checking with no value preparation or casting. Only types directly supported by the auto adaption will pass the checks. From early tests 10-20% slower than 'none', but getting proper type checks on python side sounds like a good _default_.
- 'broad'
  Mode mimicking django's relatively broad input values with type casts. This mode exists, because `get_db_prep_save` does alot more at the final adaption stage, which is not needed with COPY/binary, but doubles the runtime.
- 'django'
  Mode using django's `get_db_prep_save` methods. This is actually the trickiest, as some final value adaptions are wrong for COPY/binary and thus need an additional cast step. Runs 4-5 times slower than 'strict'.

On API side the user then can decide, which preparation mode to use:
```python
def copy_create(
    qs: models.QuerySet,
    objs: Sequence[models.Model],
    returns: bool = False,
    from_dict: bool = False,
    default_prepare: str = 'strict',
    prepare_overrides: Optional[Dict[str, Union[str, Callable[..., Any]]]] = None
) -> Any: ...
```
where _default_prepare_ denotes the default mode applied to all fields, except those overridden in _prepare_overrides_.

**Comment by jerch** [comment]
Runtime for different prepare modes with _FieldUpdate_ model:

```python
>>> from fast_update.copy3 import copy_create
>>> from postgres_tests.tests import CU_EXAMPLE
>>> o = [FieldUpdate(**CU_EXAMPLE) for _ in range(10000)]
>>> _=timer(lambda:copy_create(FieldUpdate.objects.all(), o, default_prepare='none'))
0.23005366325378418
>>> _=timer(lambda:copy_create(FieldUpdate.objects.all(), o, default_prepare='strict'))
0.32703518867492676
>>> _=timer(lambda:copy_create(FieldUpdate.objects.all(), o, default_prepare='django'))
0.679027795791626
>>> _=timer(lambda:FieldUpdate.objects.bulk_create(o))
1.7566735744476318
```
Since the model has a JSON field, we can further speed it up by using `orjson`:
```python
>>> import orjson
>>> _=timer(lambda:copy_create(FieldUpdate.objects.all(), o, default_prepare='strict', json_dumps=orjson.dumps))
0.19280123710632324
>>> _=timer(lambda:copy_create(FieldUpdate.objects.all(), o, default_prepare='none', json_dumps=orjson.dumps))
0.1364123821258545
```

**Comment by jerch** [comment]
There is an issue in psycopg-binary with COPY/binary - https://github.com/psycopg/psycopg/discussions/1147, which may cause an OOM error on postgres side.

Interestingly only wrong array values seem to be affected. Since we can work around the wrong array values on python side, an interim fix will be to provide 2 versions of array_factory, one relying on psycopg only and a second one doing another balance & empty list check beforehand. The second one can later be removed once the fix got widely rolled out.
